cli/src/commands/extensions/permissions.rs

```diff
@@ -1,29 +1,53 @@
 use std::path::PathBuf;
 
+use anyhow::{anyhow, Result};
 use deno_runtime::permissions::PermissionsOptions;
 use serde::de::Error as _;
 use serde::{Deserialize, Deserializer, Serialize};
 
+/// Resource permissions.
+#[derive(Serialize, Deserialize, Clone, Default, Debug)]
+pub struct Permission(Option<Vec<String>>);
+
+impl Permission {
+    // XXX In Deno, `Some(vec![])` actually means "allow all". We never
+    // want that, so we manually convert those instances into `None` in the
+    // getter methods below. We need to make sure to always go through these when
+    // constructing a `PermissionsOptions` object.
+    pub fn get(&self) -> Option<&Vec<String>> {
+        self.0.as_ref().and_then(|v| if v.is_empty() { None } else { Some(v) })
+    }
+
+    /// Check if access to resource is permitted.
+    pub fn validate(&self, resource: &String, resource_type: &str) -> Result<()> {
+        if self.get().map_or(false, |allowed| allowed.contains(resource)) {
+            Ok(())
+        } else {
+            Err(anyhow!("Requires {resource_type} access to {resource:?}"))
+        }
+    }
```

Comment:
I pass the `resource_type` in here just for the error, but I'd love to hear about ideas how this could be automatically inferred from the `Permission` struct.

I've considered adding like a trait bound with a phantomdata on it, then specify permissions as `Permission<Net>` where `Net` has a function to return `"net"`, but that seems like **a lot** of additional boilerplate.

Reply:
I think it is not necessary to have this right now, but going forward if we realize we want some form of validation having specialized permission types could be an option there. Those could implement some trait with default methods. It is also quite a bit of boilerplate, so I'd say we're still in the same league.

Another option could be to just add an enum field to the permissions object, which maybe could be a bit less heavy handed on boilerplate than using phantom data but is also non-zero cost.